src/codegen/query.cpp

```diff
@@ -70,44 +92,42 @@ void Query::Execute(std::unique_ptr<executor::ExecutorContext> executor_context,
 
   // Time initialization
   if (stats != nullptr) {
-    timer.Stop();
-    stats->init_ms = timer.GetDuration();
-    timer.Reset();
-    timer.Start();
+    timer->Stop();
+    stats->init_ms = timer->GetDuration();
+    timer->Reset();
+    timer->Start();
   }
 
-  // Execute the query!
-  LOG_TRACE("Calling query's plan() ...");
-  try {
-    plan_func_(param);
-  } catch (...) {
-    // Cleanup if an exception is encountered
+  auto on_execute_stages_complete = [this, param, context, timer, stats, on_complete] {
+    // Timer plan execution
+    if (stats != nullptr) {
+      timer->Stop();
+      stats->plan_ms = timer->GetDuration();
+      timer->Reset();
+      timer->Start();
+    }
+
+    // Clean up
+    LOG_TRACE("Calling query's tearDown() ...");
     tear_down_func_(param);
-    throw;
-  }
 
-  // Timer plan execution
-  if (stats != nullptr) {
-    timer.Stop();
-    stats->plan_ms = timer.GetDuration();
-    timer.Reset();
-    timer.Start();
-  }
+    // No need to cleanup if we get an exception while cleaning up...
+    if (stats != nullptr) {
+      timer->Stop();
+      stats->tear_down_ms = timer->GetDuration();
+    }
 
-  // Clean up
-  LOG_TRACE("Calling query's tearDown() ...");
-  tear_down_func_(param);
+    executor::ExecutionResult result;
+    result.m_result = ResultType::SUCCESS;
+    result.m_processed = context->num_processed;
+    on_complete(result);
 
-  // No need to cleanup if we get an exception while cleaning up...
-  if (stats != nullptr) {
-    timer.Stop();
-    stats->tear_down_ms = timer.GetDuration();
-  }
+    delete[] param;
```

Comment:
Reason going back to non-smart pointers?

Reply:
These objects need to have a lifetime throughout the execution of the query.

Now that the query is split into separate stage functions, all these functions need to have access to the objects. This is done by lambda capture.

However in C++11 we don't have move-capture. A broader problem is that C++ doesn't have non-copy-able `std::function`s (`folly::Function` does that).

Therefore, we just go back to manually destroying the objects in the last function.